src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptions.java

```diff
@@ -596,6 +589,96 @@ public boolean apply(Object input) {
     }
 
     /**
+     * Tests that emulates a frozen elected master node that unfreezes and pushes his cluster state to other nodes
+     * that already are following another elected master node. These nodes should reject this cluster state and prevent
+     * them from following the stale master.
+     */
+    @Test
+    public void testStaleMasterNotHijackingMajority() throws Exception {
+        // TODO: on mac OS multicast threads are shared between nodes and we therefore we can't simulate GC and stop pinging for just one node
+        // find a way to block thread creation in the generic thread pool to avoid this.
+        // 3 node cluster with unicast discovery and minimum_master_nodes set to 2:
+        List<String> nodes = startUnicastCluster(3, null, 2);
+
+        // Save the current master node as old master node, because that node will get frozen
+        final String oldMasterNode = internalCluster().getMasterName();
+        for (String node : nodes) {
+            ensureStableCluster(3, node);
+        }
+        assertMaster(oldMasterNode, nodes);
+
+        // Simulating a painful gc by suspending all threads for a long time on the current elected master node.
+        SingleNodeDisruption masterNodeDisruption = new LongGCDisruption(getRandom(), oldMasterNode);
+
+        // Save the majority side
+        final List<String> majoritySide = new ArrayList<>(nodes);
+        majoritySide.remove(oldMasterNode);
+
+        // Keep track of the masters that appear in the cluster state on both nodes on the majority side,
+        // only new_master may appear in here.
+        final Map<String, Tuple<String, String>> masters = Collections.synchronizedMap(new HashMap<String, Tuple<String, String>>());
+        for (String node : majoritySide) {
+            internalCluster().getInstance(ClusterService.class, node).add(new ClusterStateListener() {
+                @Override
+                public void clusterChanged(ClusterChangedEvent event) {
+                    DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+                    DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+                    if (!Objects.equals(previousMaster, currentMaster)) {
+                        String previousMasterNodeName = previousMaster != null ? previousMaster.name() : null;
+                        String currentMasterNodeName = currentMaster != null ? currentMaster.name() : null;
+                        masters.put(currentMasterNodeName, new Tuple<>(previousMasterNodeName, currentMasterNodeName));
```

Comment:
The `masters ` map keeps track of the unique master changes, so that is why I used `currentMasterNodeName` (not all the `node` vars are masters), otherwise we can't track all changes. If I use `node` as key then I would change the `masters` to have a list of tuples that should have size 2 and have a null and newMasterNode as entries respectively.